test/deploy.py

```diff
@@ -0,0 +1,569 @@
+
+        if "vm_name" in model:
+            vm_name = model["vm_name"]
+            if self.client.instances.exists(vm_name):
+                print("Deleting VM {}".format(vm_name))
+                vm = self.client.virtual_machines.get(vm_name)
+                vm.stop(wait=True)
+                vm.delete(wait=True)
+
+        if "profile" in model:
+            profile_name = model["profile"]
+            if self.client.profiles.exists(profile_name):
+                print("Deleting VM Profile {}".format(vm_name))
+                self.client.profiles.get(profile_name).delete()
+
+        if "storage_pool" in model:
+            pool_name = model["storage_pool"]
+            if self.client.storage_pools.exists(pool_name):
+                pool = self.client.storage_pools.get(pool_name)
+
+                if "volumes" in model:
+                    volumes = model["volumes"]
+                    for volume in volumes:
+                        # Delete Volume.
+                        print(
+                            "Deleting Volume {} from Pool {}"
+                            .format(volume, pool_name)
+                        )
+                        pool.volumes.get("custom", volume).delete()
+
+                # Delete Storage pool
+                print("Deleting Pool {}".format(pool_name))
+                pool.delete()
+
+
+class DeployRunner:
+    # LXD Vars
+    deploy_tag = "oci-testing-" + _get_random_string(4)
+    complete_repo_path = "/home/ubuntu"
+    cwd = os.getcwd()
+    usr = pwd.getpwuid(os.getuid())[0]
+    modelFilePath = ""
+    model = dict()
+
+    def __init__(self) -> None:
+        # Check LXD installed on host.
+        self.check_snaps_installed()
+        # Check if current user is part of the lxd user group.
+        self.check_user_in_group()
+        # init client
+        self.client = pylxd.Client()
+        # Check if LXD is initialised.
+        self.check_lxd_initialised()
+        # File to store LXD virtual resource references.
+        self.modelFilePath = "{}/model-{}.json".format(
+            self.cwd, _get_random_string(4)
+        )
+
+    def save_model_json(self):
+        """Save vm resource dictionary to json file."""
+        with open(self.modelFilePath, "w") as model_file:
+            model_file.write(json.dumps(self.model, indent=4))
+            print("Model information exported to {}"
+                  .format(self.modelFilePath))
+
+    def check_snaps_installed(self, required_snaps: list = []):
+        """Check if snap dependencies are met."""
+        if not required_snaps:
+            required_snaps = ["lxd"]
+
+        cmd = ["snap", "list"]
+        output = subprocess.check_output(cmd).decode()
+        snaps = list(map(
+                lambda snap_entry: snap_entry.split(" ")[0],
+                output.splitlines()
+            ))
+
+        if not all(snap in snaps for snap in required_snaps):
+            raise AssertionError("Required snaps not installed: {}"
```

Comment:
I have to say that `AssertionError` is a weird choice, imo. It's supposed to by raised by `assert` only, and having it in our code doesn't feel right. Some unit testing libraries treat this exception type specially, for example. I would probably create a custom error type so that it's obvious where the problem stems from.

Reply:
Yeah, I would be lying if I say it didn't feel weird at all to me. I will switch to a custom error. Thanks.